Add OTEL messaging semantic conventions to Kafka integration

g7ed6e opened the pull request:
## Description

Add OTEL messaging semantic conventions to `Aspire.ConfluentKafka` component/integration.

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [ ] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5765)

g7ed6e committed:
Add OTEL messaging semantic conventions

Add a specific KafkaContainerBuilder matching confluentinc/confluent-local:7.7.1 requirements

Update tests/Aspire.Confluent.Kafka.Tests/MeterProviderExtensions.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Update ConfigurationSchema.json

Update src/Vendoring/README.md

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Use constants

g7ed6e commented:
@mitchdenny Could you add this package https://www.nuget.org/packages/OpenTelemetry.Instrumentation.ConfluentKafka/0.1.0-alpha.2 ?

mitchdenny commented:
> @mitchdenny Could you add this package https://www.nuget.org/packages/OpenTelemetry.Instrumentation.ConfluentKafka/0.1.0-alpha.2 ?

Its importing now.

g7ed6e commented:
/azp run

davidfowl commented on src/Aspire.Hosting/Aspire.Hosting.csproj:
```diff
@@ -47,6 +47,7 @@
     <PackageReference Include="KubernetesClient" />
     <PackageReference Include="Microsoft.Extensions.Hosting" />
     <PackageReference Include="Polly.Core" />
+    <PackageReference Include="System.Text.Json" />
```
?

g7ed6e replied:
@davidfowl In VS2022 preview it looks like NuGetAudit is enabled by default. Locally I had to update STJ to latest version.
I updated the PR dropping this PackageReference and keeping a `<NuGetAudit>false</NuGetAudit>` MSBuild property in my local Directory.Build.props.

g7ed6e commented:
@davidfowl / @eerhardt We now have 2 metrics sources.
![image](https://github.com/user-attachments/assets/9fe79526-5960-4074-b508-a3539102c846)

Should we still emit both ? Should we add an AppContext.Switch ?

eerhardt commented:
> Should we still emit both ? Should we add an AppContext.Switch ?

I think we should start deprecating the existing one. Either removing it entirely now, or putting it behind an AppContext switch first, and then remove it later.

eerhardt commented:
Can you update

https://github.com/dotnet/aspire/blob/62f8cf792deee900b8afd04c8e9e278c7dc6fa8a/src/Components/Telemetry.md?plain=1#L88-L105

eerhardt commented on Directory.Packages.props:
```diff
@@ -139,6 +139,7 @@
     <PackageVersion Include="OpenTelemetry.Instrumentation.Http" Version="1.9.0" />
     <PackageVersion Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.9.0" />
     <PackageVersion Include="OpenTelemetry.Instrumentation.Runtime" Version="1.9.0" />
+    <PackageVersion Include="OpenTelemetry.Instrumentation.ConfluentKafka" Version="0.1.0-alpha.2" />
```
What's the plan for shipping this library stable? We won't be able to ship `Aspire.Confluent.Kafka` stable again until this package goes stable.

Kielek replied:
https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-spans.md and https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/messaging-metrics.md needs to be stable first. Than we can think about making this package stable.

eerhardt replied:
I'm assuming that isn't going to happen this calendar year, is that correct?

If that is the case, maybe we need to think about "vendoring" this code and compiling it into the Aspire.Confluent.Kafka library like we do with SqlServer and Redis here:

https://github.com/dotnet/aspire/tree/main/src/Vendoring

davidfowl replied:
So we're going to need to copy the implementation @eerhardt ?

Kielek replied:
I do not found any ETA for this. My feeling is more likely - early next year.

eerhardt replied:
> So we're going to need to copy the implementation @eerhardt ?

If we want to continue shipping "stable" versions of the Kafka client integration library.

eerhardt commented:
Can you add some tests?

davidfowl commented:
@g7ed6e we're marching towards the date for Aspire 9 shutdown, are you interested in having this support land in 9?

g7ed6e commented:
> Can you update
> 
> https://github.com/dotnet/aspire/blob/62f8cf792deee900b8afd04c8e9e278c7dc6fa8a/src/Components/Telemetry.md?plain=1#L88-L105

done

g7ed6e commented:
> > Should we still emit both ? Should we add an AppContext.Switch ?
> 
> I think we should start deprecating the existing one. Either removing it entirely now, or putting it behind an AppContext switch first, and then remove it later.

done

g7ed6e commented on tests/Aspire.Confluent.Kafka.Tests/.cs:
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
it contains a ModuleInitializer that set the AppContext switch. How would you name it ? Program.cs ?

g7ed6e commented on tests/Aspire.Confluent.Kafka.Tests/.cs:
```diff
@@ -0,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
```
@eerhardt These tests would require a Kafka container running (this is how I wrote the tests in the OTEL repo). Is this what you would like to see here ?

g7ed6e replied:
Please see here https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/test/OpenTelemetry.Instrumentation.ConfluentKafka.Tests/HostedTracingAndMeteringTests.cs

g7ed6e replied:
Hum I did not noticed yet  we were pulling TestContainers. Let's write integration tests :)

g7ed6e replied:
@eerhardt / @mitchdenny `TestContainers.Kafka` default image is confluentinc/cp-kafka:6.1.9 can you please:
- add confluentinc/cp-kafka:6.1.9 to netaspireci.azurecr.io
- add TestContainers.Kafka 3.10.0 to the nuget feed

g7ed6e commented:
@davidfowl I updated the PR with integration tests similar to those present in otel repo, i hope we are still able to make this land for .NET9.

CI is failing but should work once:
- `confluentinc/cp-kafka:6.1.9` is added to netaspireci.azurecr.io registry
- `TestContainers.Kafka` 3.10.0 to nuget feed

mitchdenny commented:
Testcontainers package should be in there in a few moments. I'm working on getting my access sorted for the container image.

eerhardt commented:
/azp run

eerhardt commented:
> confluentinc/cp-kafka:6.1.9 is added to netaspireci.azurecr.io registry

We shouldn't be using this container for tests. Instead we should use the Kafka container we reference in our Hosting package - 

https://github.com/dotnet/aspire/blob/4e7e0de000bb83f824cca2773892a22b5da89245/src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs#L12-L15

You can override the TestContainer's image like the following:

https://github.com/dotnet/aspire/blob/4e7e0de000bb83f824cca2773892a22b5da89245/tests/Aspire.RabbitMQ.Client.Tests/RabbitMQContainerFixture.cs#L34-L38

g7ed6e commented:
> > confluentinc/cp-kafka:6.1.9 is added to netaspireci.azurecr.io registry
> 
> We shouldn't be using this container for tests. Instead we should use the Kafka container we reference in our Hosting package -
> 
> https://github.com/dotnet/aspire/blob/4e7e0de000bb83f824cca2773892a22b5da89245/src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs#L12-L15
> 
> You can override the TestContainer's image like the following:
> 
> https://github.com/dotnet/aspire/blob/4e7e0de000bb83f824cca2773892a22b5da89245/tests/Aspire.RabbitMQ.Client.Tests/RabbitMQContainerFixture.cs#L34-L38

@eerhardt `confluentinc/cp-kafka:6.1.9` and `confluentinc/confluent-local` do not need the same env variables and startup script won't be compatible between the 2.
So we have 2 posibilities: 
1. drop the dependency on `TestContainers.Kafka`, pull `TestContainer` and implement our own KafakContainer and KafkaBuilder.
2. use the community provided `TestContainers.Kafka` nuget package and get benefit on its probable future upgrade to `confluentinc/confluent-local`.

One thought: As the aim of these tests is to validate that the kafka clients behave correctly we could rely on whatever Kafka implementation and should not rely on the actual kafka server version. I mean that the aspire kafka integration should be compatible with others implementation like [RedPanda](https://www.redpanda.com/home) for which a TestContainer.RedPanda [NuGet package](https://www.nuget.org/packages/Testcontainers.Redpanda) is published too.

g7ed6e commented:
@eerhardt I've updated the PR adding a specific KafkaContainerBuilder for the requirements of the confluentinc/confluent-local:7.7.1 image. The implementation is inspired by the [TestContainers go library](https://github.com/testcontainers/testcontainers-go/blob/main/modules/kafka/kafka.go) which uses a confluent/confluent-local 7.x image too. (In opposition to [the .NET one which us a confluent/cp-kafka 6.x image](https://github.com/testcontainers/testcontainers-dotnet/blob/develop/src/Testcontainers.Kafka/KafkaBuilder.cs)).

g7ed6e commented on tests/Aspire.Confluent.Kafka.Tests/.cs:
```diff
@@ -0,0 +1,16 @@
+    [ModuleInitializer]
+    public static void Initialize()
+    {
+        AppContext.SetSwitch("EnableAspire8ConfluentKafkaMetrics", true);
```
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/KafkaContainerFixture.cs:
```diff
@@ -0,0 +1,111 @@
+{
+    private sealed partial class ConfluentLocalKafkaBuilder : ContainerBuilder<ConfluentLocalKafkaBuilder, KafkaContainer, KafkaConfiguration>
+    {
+        public const string KafkaImage = "confluentinc/confluent-local:7.7.1";
```
We should reuse the consts from KafkaContainerImageTags.

https://github.com/dotnet/aspire/blob/39912824d33de54be24baf842bf9228a6a84ac33/src/Aspire.Hosting.Kafka/KafkaContainerImageTags.cs#L12-L15

That way when they are updated, they are updated in the tests as well.

You can do this by source sharing into the test csproj, like this:

https://github.com/dotnet/aspire/blob/39912824d33de54be24baf842bf9228a6a84ac33/tests/Aspire.Hosting.Keycloak.Tests/Aspire.Hosting.Keycloak.Tests.csproj#L13

g7ed6e replied:
done

eerhardt commented on tests/Aspire.Confluent.Kafka.Tests/KafkaContainerFixture.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            return base.Init()
+                .WithImage(KafkaImage)
+                .WithPortBinding(KafkaPort, true)
```
Do we need to copy all this from test containers? Can't we just override the Image like we do elsewhere?

https://github.com/dotnet/aspire/blob/39912824d33de54be24baf842bf9228a6a84ac33/tests/Aspire.RabbitMQ.Client.Tests/RabbitMQContainerFixture.cs#L36-L37

g7ed6e replied:
No we can't, that's why i first used the kafka 6.x images. 

Please notice that client side packages should not be bound to a specific version of kafka. Confluent.Kafka producer and consumer can be used with other queue system that support the kafka protocol like redpanda or Azure Event Hub.

eerhardt reviewed (commented):
Sorry for the delay @g7ed6e. This is looking really good. I just have a few comments. Once those are addressed I think we should be able to merge this.

I have added `Testcontainers.Kafka` 4.0.0 to our nuget mirror so CI should be unblocked.

g7ed6e commented on src/Components/Aspire.Confluent.Kafka/KafkaConsumerSettings.cs:
```diff
@@ -37,6 +37,14 @@ public sealed class KafkaConsumerSettings
+    /// <value>
+    /// The default value is <see langword="false"/>.
+    /// </value>
+    public bool DisableTracing { get; set; }
```
done

g7ed6e commented:
> Sorry for the delay @g7ed6e. This is looking really good. I just have a few comments. Once those are addressed I think we should be able to merge this.
> 
> 
> 
> I have added `Testcontainers.Kafka` 4.0.0 to our nuget mirror so CI should be unblocked.

@eerhardt Thanks. CI is all green now. I updated the PR according to your comments and rebased main branch.

g7ed6e commented:
@eerhardt may i ask a review ?

eerhardt reviewed (approved):
This looks great! Thanks for the great work and contribution here.

I just had one nit to follow coding conventions (use a property when used outside of the class). But I think this can be merged.

Since the PR validation hasn't run in a while, we should re-run with the current code just to make sure everything is still green.

eerhardt committed:
PR feedback. Use a property.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into feature/add-otel-confluentkafka

davidfowl commented:
@lmolkova wheh you have some time can you look
